rfcs/0031-ppc64-platforms.md

```diff
@@ -0,0 +1,86 @@
+* Bootloader tooling:
+    * the internal option *system.boot.loader.kernelFile* is currently set to the kernelTarget attribute of the host platform definition.
+      PowerPC kernels however use `zImage` as build target, yet produce a `vmlinux` file.
+      A one-line change would allow the option to default to the host platform definition's kernelFile attribute, if set.
+    * A typical bootloader for OpenPOWER systems is Petitboot. It can read a variety of bootloader configuration files, including those of GRUB 2 and SYSLINUX.
+      Using GRUB, even with *boot.loader.grub.device* = "nodev", as it is, implies building GRUB for the host architecture, which fails for ppc64le (currently because it requires soft-float).
+
+      However, the generic-extlinux-compatible bootloader module can be modified to provide options to store the configuration files in a path that Petitboot will consider and use absolute paths starting from the filesystem root, so that Petitboot can load the kernel/initrd images.
+
+Beyond a headless system, the biggest obstacle for ppc64(le) support seems to be Rust, because a lot of desktop software has it as a (direct or transitive) build dependency.
+In particular, a lot of GNOME software has a dependency on librsvg (which has parts written in Rust) via their documentation tools.
+The author of this RFC plans to create a way to build bootstrapping binaries for Rust on ppc64le (either via cross-building from x86 or from source via mrustc).
+
+Some packages, like OpenBLAS, will need target descriptions for these platforms, but many don't require changes to the actual build process.
+
+Support for Big Endian ppc64 might require changes in a few packages, though the majority of changes are assumed to be upstreamable (see [Unresolved Questions][unresolved]).
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Supporting another architecture might introduce additional maintenance overhead for nixpkgs.
+Generally this is restricted to packages that ship (JIT or AOT) compilers or do some feature selection based on architectures.
+
+If a new Rust bootstrap is created for ppc64le, as long as it isn't adopted for the other platforms, it will be a second rustc bootstrap that may cause maintenance effort.
+
+If a build infrastructure of ppc64le systems should be established, it will cause maintenance effort as well as costs for either procurement of hardware, or use of public cloud services (e. g. IntegriCloud), unless an organisation is willing to sponsor build machines. 
+
+# Alternatives
+[alternatives]: #alternatives
+
+For ppc64(le) in general:
+
+* keep the already-merged stdenv/bootstrapping code, maintain package changes as an overlay
+* revert stdenv/bootstrapping support, maintain ppc64le as a nixpkgs fork
+* not have ppc64le support in any NixOS-based distribution (barely an alternative)
+
+# Unresolved questions
+[unresolved]: #unresolved-questions
+
+Should the nixpkgs/NixOS project host bootstrap files for ppc64le?
```

Comment:
The informal NixOS hardware team can provide some file hosting if necessary but it can't look official.